libdirections-onboard/src/main/java/com/mapbox/navigation/route/onboard/MapboxOnboardRouter.kt

```diff
@@ -121,23 +125,29 @@ class MapboxOnboardRouter : Router {
 
     private fun retrieveRoute(url: String, callback: Router.Callback) {
         mainJobControl.scope.launch {
-            val routerResult = withContext(ThreadController.IODispatcher) {
-                navigatorNative.getRoute(url)
-            }
-
-            val routes: List<DirectionsRoute> = try {
-                DirectionsResponse.fromJson(routerResult.json).routes()
-            } catch (e: RuntimeException) {
-                emptyList()
-            }
-
-            when {
-                !routes.isNullOrEmpty() -> callback.onResponse(routes)
-                else -> callback.onFailure(NavigationException(generateErrorMessage(routerResult.json)))
+            try {
+                val routerResult = getRoute(url)
+
+                val routes: List<DirectionsRoute> = try {
```

Comment:
Do we want to do the serialization in the main thread? cc @korshaknn

Reply:
We definitely shouldn't, nice catch!